src/embassy/time_driver.rs

```diff
@@ -0,0 +1,394 @@
+        #[cfg(feature = "time-driver-tim14")]
+        {
+            unsafe {
+                cortex_m::peripheral::NVIC::unmask(Interrupt::TIMER14);
+            }
+        }
+
+        timer.ctl0().modify(|_, w| w.cen().enabled());
+    }
+
+    fn get_alarm<'a>(&'a self, cs: CriticalSection<'a>, alarm: AlarmHandle) -> &'a AlarmState {
+        // safety: we're allowed to assume the AlarmState is created by us, and
+        // we never create one that's out of bounds.
+        unsafe { self.alarms.borrow(cs).get_unchecked(alarm.id() as usize) }
+    }
+
+    fn trigger_alarm(&self, n: usize, cs: CriticalSection) {
+        let alarm = &self.alarms.borrow(cs)[n];
+        alarm.timestamp.set(u64::MAX);
+        // Call after clearing alarm, so the callback can set another alarm.
+
+        // The callback must have been set by `allocate_alarm` so this shouldn't panic.
+        let callback = alarm.callback.get().unwrap();
+        let f: fn(*mut ()) = callback.0;
+        f(callback.1);
+    }
+
+    fn on_interrupt(&self) {
+        let r = timer();
+
+        critical_section::with(|cs| {
+            let status = r.intf().read();
+            let enabled = r.dmainten().read();
+
+            // Clear all interrupt flags. Bits in SR are "write 0 to clear", so write the bitwise NOT.
+            // Other approaches such as writing all zeros, or RMWing won't work, they can
+            // miss interrupts.
+            r.intf().write(|w| unsafe { w.bits(!status.bits()) });
+
+            if status.upif().bit() {
+                self.next_period()
+            }
+
+            if status.ch0if().bit() {
+                self.next_period()
+            }
+
+            for n in 0..ALARM_COUNT {
+                // Bit 0 indicates Update CC interrupt, bit 1 corresponds to reserved Channel 0
+                let channel = n + 2;
+                if ((enabled.bits() >> channel) & 1 != 0) && ((status.bits() >> channel) & 1 != 0) {
+                    self.trigger_alarm(n, cs);
+                }
+            }
+        })
+    }
+
+    fn next_period(&self) {
+        let r = timer();
+
+        // We only modify the period from the timer interrupt, so we know this can't race.
+        let period = self.period.load(Ordering::Relaxed) + 1;
+        self.period.store(period, Ordering::Relaxed);
+        let t = (period as u64) << 15;
+
+        critical_section::with(move |cs| {
+            for n in 0..ALARM_COUNT {
+                let alarm = &self.alarms.borrow(cs)[n];
+                let at = alarm.timestamp.get();
+
+                if at < t + 0xc000 {
+                    // just enable it. `set_alarm` has already set the correct CCR val.
+                    r.set_chn_ie(n + 1, true);
+                }
+            }
+        })
+    }
+}
+
+impl Driver for EmbassyTimeDriver {
+    fn now(&self) -> u64 {
+        let r = timer();
+        let period = self.period.load(Ordering::Relaxed);
+        compiler_fence(Ordering::Acquire);
```

Comment:
Would using `Ordering::Acquire` for the period load operation on the previous line have an equivalent effect?